README.rst

```diff
@@ -13,150 +12,145 @@ smart_open -- utils for streaming large files
-  ...     print line
+  >>> for line in smart_open('webhdfs://host:port/user/hadoop/my_file.txt'):
+  ...     print(line)
 
   >>> # stream content *into* S3 (write mode):
-  >>> with smart_open.smart_open('s3://mybucket/mykey.txt', 'wb') as fout:
-  ...     for line in ['first line', 'second line', 'third line']:
-  ...          fout.write(line + '\n')
+  >>> with smart_open('s3://mybucket/mykey.txt', 'wb') as fout:
+  ...     for line in [b'first line\n', b'second line\n', b'third line\n']:
+  ...          fout.write(line)
 
   >>> # stream content *into* HDFS (write mode):
-  >>> with smart_open.smart_open('hdfs://host:port/user/hadoop/my_file.txt', 'wb') as fout:
-  ...     for line in ['first line', 'second line', 'third line']:
-  ...          fout.write(line + '\n')
+  >>> with smart_open('hdfs://host:port/user/hadoop/my_file.txt', 'wb') as fout:
+  ...     for line in [b'first line\n', b'second line\n', b'third line\n']:
+  ...          fout.write(line)
 
   >>> # stream content *into* WebHDFS (write mode):
-  >>> with smart_open.smart_open('webhdfs://host:port/user/hadoop/my_file.txt', 'wb') as fout:
-  ...     for line in ['first line', 'second line', 'third line']:
-  ...          fout.write(line + '\n')
+  >>> with smart_open('webhdfs://host:port/user/hadoop/my_file.txt', 'wb') as fout:
+  ...     for line in [b'first line\n', b'second line\n', b'third line\n']:
+  ...          fout.write(line)
 
-  >>> # stream from/to local compressed files:
-  >>> for line in smart_open.smart_open('./foo.txt.gz'):
-  ...    print line
+  >>> # stream using a completely custom s3 server, like s3proxy:
+  >>> for line in smart_open('s3u://user:secret@host:port@mybucket/mykey.txt', 'rb'):
+  ...    print(line.decode('utf8'))
 
-  >>> with smart_open.smart_open('/home/radim/foo.txt.bz2', 'wb') as fout:
-  ...    fout.write("some content\n")
+  >>> # you can also use a boto.s3.key.Key instance directly:
+  >>> key = boto.connect_s3().get_bucket("my_bucket").get_key("my_key")
+  >>> with smart_open(key, 'rb') as fin:
+  ...     for line in fin:
+  ...         print(line.decode('utf8'))
 
-Since going over all (or select) keys in an S3 bucket is a very common operation,
-there's also an extra method ``smart_open.s3_iter_bucket()`` that does this efficiently,
-**processing the bucket keys in parallel** (using multiprocessing):
 
-.. code-block:: python
+Why?
+----
 
-  >>> # get all JSON files under "mybucket/foo/"
-  >>> bucket = boto.connect_s3().get_bucket('mybucket')
-  >>> for key, content in s3_iter_bucket(bucket, prefix='foo/', accept_key=lambda key: key.endswith('.json')):
-  ...     print key, len(content)
+Working with large S3 files using Amazon's default Python library, `boto <http://docs.pythonboto.org/en/latest/>`_ and `boto3 <https://boto3.readthedocs.io/en/latest/>`_, is a pain. Its ``key.set_contents_from_string()`` and ``key.get_contents_as_string()`` methods only work for small files (loaded in RAM, no streaming).
```

Comment:
Need to check, how this rendered on PyPI (as I remember, PyPI have really "strict" parser and readme render can be easily broken)

Reply:
I don't think I made any changes here (just moved the sections around).

Reply:
Checked, rendering works https://test.pypi.org/project/smart-open/

Reply:
Great. I didn't know about test.pypi.org, nice find!